python_modules/dagster/dagster/_core/storage/upath_io_manager.py

```diff
@@ -96,16 +97,58 @@ def _get_paths_for_partitions(
                 "but the asset is not partitioned"
             )
 
+        def _formatted_multipartitioned_path(partition_key: MultiPartitionKey) -> str:
+            ordered_dimension_keys = [
+                key[1]
+                for key in sorted(partition_key.keys_by_dimension.items(), key=lambda x: x[0])
+            ]
+            return "/".join(ordered_dimension_keys)
+
+        formatted_partition_keys = [
+            _formatted_multipartitioned_path(pk) if isinstance(pk, MultiPartitionKey) else pk
+            for pk in context.asset_partition_keys
+        ]
+
+        asset_path = self._get_path_without_extension(context)
+        return {
+            partition_key: self._with_extension(asset_path / partition_key)
+            for partition_key in formatted_partition_keys
+        }
+
+    def _get_multipartition_backcompat_paths(
+        self, context: Union[InputContext, OutputContext]
+    ) -> UPath:
+        if not context.has_asset_partitions:
+            raise TypeError(
+                f"Detected {context.dagster_type.typing_type} input type "
+                "but the asset is not partitioned"
+            )
+
         partition_keys = context.asset_partition_keys
+
         asset_path = self._get_path_without_extension(context)
         return {
             partition_key: self._with_extension(asset_path / partition_key)
             for partition_key in partition_keys
+            if isinstance(partition_key, MultiPartitionKey)
         }
 
-    def _load_single_input(self, path: UPath, context: InputContext) -> Any:
+    def _load_single_input(
+        self, path: UPath, context: InputContext, backcompat_path: Optional[UPath] = None
+    ) -> Any:
         context.log.debug(f"Loading file from: {path}")
-        obj = self.load_from_path(context=context, path=path)
+        try:
+            obj = self.load_from_path(context=context, path=path)
+        except FileNotFoundError as e:
+            if backcompat_path is not None:
+                context.log.debug(
```

Comment:
I think that we'd want to surface the backcompat path as little as possible in logs and exceptions, to avoid confusing people. Can we make it so that they only see a log if we successfully load from the backcompat path? And that, if we fail to load from either, the path that comes up in the stack trace is the non-backcompat path?

Reply:
Thanks for the feedback--this makes a lot of sense. This PR has now been updated to follow this behavior.

The error we now raise if no file exists at either path is the non-backcompat path. We only display the log message now if the file is found at the backcompat path.